enhancements/cloud-integration/azure/azure-workload-identity.md

```diff
@@ -0,0 +1,401 @@
+```
+
+#### Credentials secret
+
+OpenShift operators currently obtain their long-lived credentials from a config secret with the following format:
+
+```yaml
+apiVersion: v1
+data:
+  azure_client_id: <client id>
+  azure_client_secret: <client secret>
+  azure_region: <region>
+  azure_resource_prefix: <resource group prefix eg. "abutcher-az-t68n4">
+  azure_resourcegroup: <resource group eg. "abutcher-az-t68n4-rg">
+  azure_subscription_id: <subscription id>
+  azure_tenant_id: <tenant id>
+kind: Secret
+type: Opaque
+```
+
+We propose that when utilizing Azure Workload Identity, the credentials secret will contain an `azure_client_id` that is the `clientID` of the Managed Identity provisioned by `ccotcl` for the operator. The `azure_client_secret` key will be absent and instead we can provide the path to the mounted `ServiceAccount` token as an `azure_federated_token_file` key; the path to the mounted token is well
+known and is specified in the operator deployment.
+
+The resource group in which the installer will create infrastructure will not be known when these secrets are generated by `ccoctl` ahead of installation and operators which rely on `azure_resourcegroup` and `azure_resource_prefix` such as the
+[image-registry](https://github.com/openshift/cluster-image-registry-operator/blob/8556fd48027f89e19daad36e280b60eb93d012d4/pkg/storage/azure/azure.go#L95-L100) should obtain the resource group details from the cluster `Infrastructure` object instead.
+
+```yaml
+apiVersion: v1
+data:
+  azure_client_id: <client id>
+  azure_federated_token_file: <path to mounted service account token, eg. "/var/run/secrets/openshift/serviceaccount/token">
+  azure_region: <region>
+  azure_subscription_id: <subscription id>
+  azure_tenant_id: <tenant id>
+kind: Secret
+type: Opaque
+```
+
+#### Creating workload identity clients in operators
+
+In order to create Azure clients which utilize a `ClientAssertionCredential`, operators must update to version `>= v1.2.0` of the azidentity package within [azure-sdk-for-go](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity@v1.2.0). Ahead of this work, due to the [end of life
+announcement](https://techcommunity.microsoft.com/t5/microsoft-entra-azure-ad-blog/microsoft-entra-change-announcements-september-2022-train/ba-p/2967454) of the Azure Active Directory Authentication Library (ADAL), PRs (ex. [openshift/cluster-ingress-operator](https://github.com/openshift/cluster-ingress-operator/pull/846)) have been opened for operators to migrate to creating clients via
+azidentity which are converted into an authorizer for use with v1 clients. Once these changes have been made, we propose that OpenShift operators continue to utilize a config secret to obtain authentication details as described in the previous section but create workload identity clients when the `azure_client_secret` is absent AND/OR when  `azure_federated_token_file` fields are found in the
```

Comment:
This "AND/OR" doesn't gel with the code sample, which is [just using "AND"](https://github.com/openshift/enhancements/pull/1301/files#diff-e2f5ffd1d09d902828e72de02ebe2f836c5b4226bde707af47d416ed33247e82R185-R186). Which is the officially supported way?

Reply:
This code is going to be brought by our implementation. I'm inclined to say that we should look for both `azure_client_secret` being absent AND `azure_federated_token_file` being set, however, we will probably already know the file location [from the deployments](https://github.com/openshift/enhancements/pull/1301/files#r1044560583). I'll update the code sample as well.